NEWS.md

```diff
@@ -550,6 +550,8 @@
 
 53. `as.data.frame(DT, row.names=)` no longer silently ignores `row.names`, [#5319](https://github.com/Rdatatable/data.table/issues/5319). Thanks to @dereckdemezquita for the fix and PR, and @ben-schwen for guidance.
 
+54. `X[Y, on="k"]` and `merge(Y, X, by="k")` as well as `merge(X, Y, by="k")` returned a wrongly keyed `data.table` if `X` is keyed by `k` and `k` is a `factor`, whose levels are not sorted, in `X` and `character` in `Y`, [#5361](https://github.com/Rdatatable/data.table/issues/5361). Thanks to @gbrunick for the report and Benjamin Schwendinger for the fix.
```

Comment:
I've read this news item 3 times and I think I get it but I'm finding myself reading it a 4th and now 5th time, and now a 6th and 7th time after writing this comment.

Would it read better as follows?

> Joining a factor column containing unsorted levels to a character column would return an invalidly keyed `data.table`. This could lead to a subsequent error/warning or silent incorrect results [TBC]. Thanks to ...

The thought process I went through :

1. A phrase such as "merge(Y, X, by='k') as well as merge(X, Y, by='k')" is one level too detailed. Although technically more correct, it makes the news item harder to read quickly.  It's hard to explain, but I'll try ... what I see is X[Y something] and merge(something) already conveying that the bug applies to both X[Y] and merge() and the fact that it doesn't matter which way around X and Y are passed to merge is "one level deeper" in my mind which distracts from the more important actual bug which is actually related to unsorted factor levels. So put unsorted factor levels first. And then after a few revisions typing this comment I realized we can just replace all of the first 52 character with one word: "joining". That is "better" because it's shorter and more direct. But just being shorter isn't always better. So I don't know how to explain it. But I feel that news items are worth worrying about to this level as they are user facing.

2. By "wrongly keyed", I am thinking from a user perspective and wondering does it lead to a warning/error or does it lead to silent incorrect results. The latter is even more critical and concerning than the former. Language conveying severity level could be standardized and added to news items standards in the contributing guide.  In this case I suspect that an invalid key is silently created which leads to silent incorrect results? If that's the case then that needs to be stated in the news item so the users know it's quite a serious bug.  Should a principal for news items be that they should convey the worst case impact?

3. If I've understood correctly so far and it's a severe bug (for the reason of silent incorrect results) then perhaps a rule should be that the news item should always exhibit an example in such cases.

Sad and surprised to see such a bug. Reading the original news items for an 8th time, I see it's only when X is keyed by the factor column, so revising the suggestion to :

> If the first column of a key is a factor column containing unsorted levels and it is joined to a character column, then an invalidly keyed result would be produced. This could lead to a subsequent error/warning or silent incorrect results [TBC]. Thanks to ...

One question I guess a user would have after reading this news item is does `setkey()` sort the levels in a factor column or leave them unsorted?  In other words, they will be wondering how easily a key column with unsorted levels can arise. Maybe we can address that up front in the news item, and doing so should somehow be added to contributing guide for news items, @MichaelChirico.

Reply:
`setkey()` does not sort the `levels` in a `factor`, but sorts the vector according to these `levels`. This is IMO the correct behavior, since, otherwise we would probably have to deal with `ordered` and other subclasses accordingly.